### ISSUE TITLE
Segfault on CPU platform

### ISSUE THREAD

**Issue by jchodera** [pandegroup/openmm#781]
The following code segfaults on the `CPU` platform for my MacBook Air running OS X 10.9.5:
```
#!/usr/bin/env python

from simtk import openmm
from simtk import unit

from openmmtools import testsystems

testsystem = testsystems.LennardJonesGrid()

temperature = 300.0 * unit.kelvin
collision_rate = 5.0 / unit.picoseconds
timestep = 2.0 * unit.femtoseconds
platform_name = 'CPU'
niterations = 100
nsteps = 50

platform = openmm.Platform.getPlatformByName(platform_name)

integrator = openmm.LangevinIntegrator(temperature, collision_rate, timestep)
context = openmm.Context(testsystem.system, integrator, platform)
context.setPositions(testsystem.positions)

state = context.getState(getEnergy=True)
print "initial               : potential = %.3f kcal/mol" % (state.getPotentialEnergy() / unit.kilocalories_per_mole)

for iteration in range(niterations):
    integrator.step(nsteps)
    state = context.getState(getEnergy=True)
    print "iteration %5d / %5d : potential = %.3f kcal/mol" % (iteration, niterations, state.getPotentialEnergy() / unit.kilocalories_per_mole)
```
The [`openmmtools`](https://github.com/choderalab/openmmtools) package can be installed via:
```
conda install -c https://conda.binstar.org/omnia openmmtools
```
The system being simulated is a [grid of Lennard-Jones atoms created with `LennardJonesGrid`](https://github.com/choderalab/openmmtools/blob/master/openmmtools/testsystems.py#L1439-L1509).

A related issue was causing segfaults on the CPU platform on travis.

**Comment by jchodera** [comment]
The `Reference` platform behavior is also strange, sometimes producing a string of the same energies, sometimes producing `NaN`.

This is a tightly packed grid, so it could be that large forces cause instabilities, but the code should never `segfault`.  Returning an `Exception` would be much more desirable.

**Comment by jchodera** [comment]
Adding the line 
```
openmm.LocalEnergyMinimizer.minimize(context) 
```
prior to integration fixes the problem most of the time.

**Comment by peastman** [comment]
Could you run it in lldb and see where the segfault is happening?  Once nans start appearing it calculations, it's very difficult to guarantee the code will never segfault.

**Comment by jchodera** [comment]
```
[LSKI1497:~/code/omnia/openmmtools.jchodera] choderaj% lldb python segfault-test-case.py
(lldb) target create "python"
Current executable set to 'python' (x86_64).
(lldb) settings set -- target.run-args  "segfault-test-case.py"
(lldb) run
Process 6172 launched: '/Users/choderaj/anaconda/bin/python' (x86_64)
initial               : potential = 349574426.255 kcal/mol
iteration     0 /   100 : potential = 20595103441.998 kcal/mol
Process 6172 stopped
* thread #1: tid = 0xdd6b77, 0x0000000103b8d699 libOpenMMCPU.dylib`OpenMM::CpuNeighborList::computeNeighborList(int, OpenMM::AlignedArray<float> const&, std::__1::vector<std::__1::set<int, std::__1::less<int>, std::__1::allocator<int> >, std::__1::allocator<std::__1::set<int, std::__1::less<int>, std::__1::allocator<int> > > > const&, float const*, bool, float, OpenMM::ThreadPool&) + 1145, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xfffffff50472b2c0)
    frame #0: 0x0000000103b8d699 libOpenMMCPU.dylib`OpenMM::CpuNeighborList::computeNeighborList(int, OpenMM::AlignedArray<float> const&, std::__1::vector<std::__1::set<int, std::__1::less<int>, std::__1::allocator<int> >, std::__1::allocator<std::__1::set<int, std::__1::less<int>, std::__1::allocator<int> > > > const&, float const*, bool, float, OpenMM::ThreadPool&) + 1145
libOpenMMCPU.dylib`OpenMM::CpuNeighborList::computeNeighborList(int, OpenMM::AlignedArray<float> const&, std::__1::vector<std::__1::set<int, std::__1::less<int>, std::__1::allocator<int> >, std::__1::allocator<std::__1::set<int, std::__1::less<int>, std::__1::allocator<int> > > > const&, float const*, bool, float, OpenMM::ThreadPool&) + 1145:
-> 0x103b8d699:  movq   (%rdi,%rdx,8), %rdx
   0x103b8d69d:  orq    $0x2, %rax
   0x103b8d6a1:  movss  (%rcx,%rax,4), %xmm0
   0x103b8d6a6:  movss  %xmm0, -0x30(%rbp)
```

**Comment by jchodera** [comment]
Is there a simple way to check if any force component is `NaN` and raise an exception at that point?  I imagine that's where things go haywire.

**Comment by jchodera** [comment]
I'm changing the default parameters in `LennardJonesGrid`, but modifying the line that constructs this object to the following will retain the old segfaulty behavior:
```
testsystem = testsystems.LennardJonesGrid(reduced_density=0.68)
```

**Comment by peastman** [comment]
The segfault is in the neighbor list building routine, so it's presumably caused by a position being nan.  It might be possible to just check every coordinate at the start of each time step, and throw an exception if anything is nan.  There would be a performance cost to that, but hopefully it wouldn't be too large.

**Comment by jchodera** [comment]
> The segfault is in the neighbor list building routine, so it's presumably caused by a position being nan

Wouldn't this have been caused by the forces being `NaN` on the previous timestep?

> It might be possible to just check every coordinate at the start of each time step, and throw an exception if anything is nan. There would be a performance cost to that, but hopefully it wouldn't be too large.

If small, probably worth it.  If large, it would be awesome if this could be an option that was maybe enabled by default, but could be switched off for "power users"?

**Comment by swails** [comment]
> If small, probably worth it. If large, it would be awesome if this could be an option that was maybe enabled by default, but could be switched off for "power users"?

I think this is backwards...  The prevailing opinion is that "more sampling is always better", and (probably too) much is made of the performance wars between codes.  In most cases, this exception will be fatal (I imagine it would not be worthwhile to attempt to programmatically figure out what went wrong and fix it given how *many* things can go wrong in our calculations).  So really all you're doing is trading `Segmentation violation` with `Exception: particle positions are NaN (you figure out why)` in the few instances where people did something "weird".  Is that really worth a "large" performance hit to "non-power users" (i.e., most people)?

I imagine it being a power user that determines it is more important to have a catchable exception with a marginally more helpful error message than a "large" performance gain.

**Comment by jchodera** [comment]
A 'segmentation fault' is not really informative to most users, and for a
library (recall that OpenMM is in fact a library and not an application),
it is simply unacceptable.

**Comment by jchodera** [comment]
> In most cases, this exception will be fatal (I imagine it would not be
worthwhile to attempt to programmatically figure out what went wrong and
fix it given how many things can go wrong in our calculations).

I don't think we need to figure out what went wrong. I think we'd like a
simple, low overhead way of preventing the code from segfaulting and taking
out whatever controlling application is running dynamics.

**Comment by peastman** [comment]
There are many ways a nan could come up, and once it appears in one place, it will quickly propagate everywhere else.  And by the time you get a nan in one place, you've probably had *something* completely absurd (huge forces, nonsensical positions, etc.) for at least one time step.  So whatever we catch, it probably won't be until after the root cause is already past.

But nans in positions are the ones most likely to directly produce a segfault, so that's the most import place to check.

I can see both sides of the issue.  I don't think a segfault is "simply unacceptable".  It is very definitely not good.  But if the alternative is throwing an exception that says, "Your simulation is messed up beyond recovery", that isn't great either.  It is, however, clearly better, since it at least gives you a chance to catch the exception and do cleanup (though most people probably won't).  So it comes down to a question of how much performance we lose.

**Comment by swails** [comment]
> A 'segmentation fault' is not really informative to most users, and for a
> library (recall that OpenMM is in fact a library and not an application),
> it is simply unacceptable.

I'm with @peastman here.  I agree it's bad, and not desirable.  But other libraries certainly *let* you segfault if you ignore their API requirements... like passing in an unallocated pointer to a place that the API mandates must be a valid array address.  Library authors typically say "if you do such-and-such, the behavior is undefined", and a segfault is perfectly permissible (think BLAS/LAPACK and the work arrays you need to provide many of the functions).

If we *can* prevent a segfault, we should.  But it's certainly not unreasonable for us to say that protecting against bad input introduces unacceptable overhead, so the resulting behavior is undefined.  It comes down to, as @peastman says, balancing the inconvenience of a segfault with the performance loss of the additional error check.

All I was trying to say before is that avoiding a "large" cost is probably more important than turning a segfault into an exception.  And if an option were to be made between an exception or a "large" performance boost, the default should benefit the majority of users.  I personally think that the majority of people would rather the performance boost than the exception in the rare instances they provide bad input.

**Comment by jchodera** [comment]
Here's my problem as the developer of an application that uses OpenMM: We
are trying to do large scale calculations on input that is often
problematic. If openmm segfaults, it takes out my whole automation
pipeline.

It's just not reasonable for a library to say 'we might return an
exception, or we might segfault, taking down your application with us---you
never know what might happen!'.  You have to find some way to trap the fuck
out of that shit if you want people to build usable applications around
your library.

**Comment by jlmaccal** [comment]
Run the problematic thing in a subprocess?

**Comment by jchodera** [comment]
Or are you suggesting OpenMM should run its kernels as a subprocess, which
is actually starting to sound like a reasonable approach to avoiding
segfaults?
On Jan 15, 2015 12:47 AM, wrote:

Yes, but that completely defeats the purpose of a library, doesn't it?

**Comment by jchodera** [comment]
Yes, but that completely defeats the purpose of a library, doesn't it?

**Comment by jchodera** [comment]
I am actually really surprised that we are debating whether segfaulting is
a reasonable return value for function calls...

**Comment by jlmaccal** [comment]
I was actually suggesting a workaround / horrible hack to not break your pipeline :)

How hard would it be to benchmark checking for NaNs in the positions / velocities / forces at the start of each timestep? I think without knowing what the performance impact is, it's hard to have a meaningful discussion.

**Comment by jlmaccal** [comment]
Actually, why not create a `CheckNaN` force that checks for NaNs in its `updateContextState`? Then we can add or not as needed.

**Comment by jchodera** [comment]
That could do the trick for us if it catches the crashes, but it may still
be awkward to explain in the manual that you should use this if you don't
want the application to randomly segfault on some data.

**Comment by swails** [comment]
I'm still hoping that a `NaN` check is cheap enough to implement a real error catch without introducing anything but insignificant overhead.  I am _not_ advocating keeping the segfault -- after all, I filed #597 which boils down to the same issue reported here.

> you should use this if you don't
> want the application to randomly segfault on some data.

This makes it sound more ridiculous than it is.  This is not a "random segfault on some data" -- it is a random segfault on some _bad_ data.  What we have here is a convenience need.  The workaround is simple -- use better data.  All I'm saying is there are worse things to me than random segfaults on bad data; a big CUDA/OpenCL performance hit being one of them (but I'm with @jlmaccal on wanting some actual performance data to inform this discussion).

But here's another idea: a segfault is actually implemented (in Linux) as the operating system sending a `SIGSEGV` signal to the process.  You can (sort of) "catch" a segfault by intercepting `SIGSEGV`.  In a Python script, try something like this:

```Python
import signal
# Register a segfault handler
def handle_segfault(sig, frame):
    """ Handle the given signal """
    # Do any cleanup or anything you need to do here
    sys.exit(1)

signal.signal(signal.SIGSEGV, handle_segfault)

# Now comes the rest of your code...
```

Note that "catching" a segfault isn't as 'safe' as catching a typical exception, since the stack or heap may have already been corrupted.  However, it will probably let you do some basic cleanup or error reporting in many cases.

Disclaimer: I'm suggesting this as a temporary band-aid, not a solution.

**Comment by jchodera** [comment]
Thanks for the SIGSEGV catching suggestion---that's an easy way to add
robustness to our applications.

My earlier point was that we are sometimes forced to operate on bad data in
large scale projects. If we had a way to check if the data was bad, your
point would be meaningful, but since the only way to check right now is to
see if it triggers a segfault. In this particular case, minimization cleans
up the data, but it isn't clear how much minimization is required in
general or whether that would clean up all such bad data.

**Comment by swails** [comment]
> My earlier point was that we are sometimes forced to operate on bad data in
> large scale projects. If we had a way to check if the data was bad, your
> point would be meaningful, but since the only way to check right now is to
> see if it triggers a segfault. In this particular case, minimization cleans
> up the data, but it isn't clear how much minimization is required in
> general or whether that would clean up all such bad data.

Why not start with a single point calculation and inspect the forces?  If they're huge (or `NaN` or `inf`), you clearly have a problem needing to be addressed (e.g., via minimization).  I actually think the issue here is that `NaN` first arises in the forces, propagates to the positions during the integration, and crashes the pairlist building.

Even when/if this issue is fixed in OpenMM, you're still going to find yourself in a situation where *something* went wrong and you will have to inspect by hand to determine the actual cause of the `NaN` or `inf`.  If you already know a list of possible causes, why not run tests _before_ the simulations rather than relying on a graceful crash?

**Comment by jchodera** [comment]
> Why not start with a single point calculation and inspect the forces? If
they're huge (or NaN or inf), you clearly have a problem needing to be
addressed (e.g., via minimization).

Checking if forces are `NaN` or `inf` at the start of integration is
certainly a good idea, and presumably wouldn't impact performance much
since it's a one-time cost at the beginning of integration and
minimization.  `NaN` or `inf`, of course, cannot be cleaned up by
minimization.

But what is the force norm cutoff above which integration would cause a
segfault?  That's hard to answer.  "huge" is not a meaningful quantitative
threshold.

We could try to be more quantitative: If the force times the timestep gives
something that we think might overflow precision, we could flag this as
problematic, but OpenMM doesn't make it easy to determine what kind of
precision model we should be using to check this.

I suspect this problem *might* be pretty general: If a user picks a
timestep for their protein system that is slightly too large, it could
cause instability and crashes. Here's what happens if I use solvated
alanine dipeptide (`testsystems.AlanineDipeptideExplicit()`) and increase
the timestep to 6 fs:

```

[LSKI1497:~/code/omnia/openmmtools.jchodera] choderaj% python
segfault-test-case.py

initial               : potential = -5892.713 kcal/mol | force norm =
116.881 kcal/mol/nm

iteration     0 /   100 : potential = -6409.174 kcal/mol

Segmentation fault
```
Sure, this is a large timestep, but that was selected to cause the segfault
quickly. If we use a timestep that is just slightly too large, it can cause
sporadic instability and random segfaults in long-running simulations.

> Even when/if this issue is fixed in OpenMM, you're still going to find
yourself in a situation where something went wrong and you will have to
inspect by hand to determine the actual cause of the NaN or inf. If you
already know a list of possible causes, why not run tests before the
simulations rather than relying on a graceful crash?

As the alanine dipeptide example suggests, there is no way to really detect
beforehand whether OpenMM will segfault.

**Comment by swails** [comment]
Does it always segfault in the pairlist builder?  Because NaN/inf checking in the pairlist building would be cheaper than doing it every step, obviously.

**Comment by jchodera** [comment]
Here's the full alanine dipeptide example:
```python
#!/usr/bin/env python

from simtk import openmm
from simtk import unit

from openmmtools import testsystems

#testsystem = testsystems.LennardJonesGrid(reduced_density=0.68)
testsystem = testsystems.AlanineDipeptideExplicit()

temperature = 300.0 * unit.kelvin
collision_rate = 5.0 / unit.picoseconds
timestep = 6.0 * unit.femtoseconds
platform_name = 'CPU'
niterations = 100
nsteps = 50

platform = openmm.Platform.getPlatformByName(platform_name)

integrator = openmm.LangevinIntegrator(temperature, collision_rate, timestep)
context = openmm.Context(testsystem.system, integrator, platform)
context.setPositions(testsystem.positions)

#openmm.LocalEnergyMinimizer.minimize(context)

state = context.getState(getEnergy=True, getForces=True)
force = state.getForces(asNumpy=True)
force_norm = unit.Quantity.sqrt(unit.Quantity.mean(force**2))

print "initial               : potential = %.3f kcal/mol | force norm = %.3f kcal/mol/nm" % (state.getPotentialEnergy() / unit.kilocalories_per_mole, force_norm / (unit.kilocalories_per_mole / unit.nanometers))

for iteration in range(niterations):
    integrator.step(nsteps)
    state = context.getState(getEnergy=True)
    print "iteration %5d / %5d : potential = %.3f kcal/mol" % (iteration, niterations, state.getPotentialEnergy() / unit.kilocalories_per_mole)
```

**Comment by jchodera** [comment]
> Does it always segfault in the pairlist builder? Because NaN/inf checking in the pairlist building would be cheaper than doing it every step, obviously.

Seems to from multiple `lldb` runs.  That could be a good idea.

**Comment by peastman** [comment]
It's impossible to guarantee that a segfault will *never* happen.  There are just too many ways they can come up.  So if it's really essential to protect your workflow against that happening, you'll need to do it at a higher level.  The best way to do that is to run each simulation in its own process (which is really easy to do with the Python multiprocessing module).

So the question is what we can do to reduce the frequency of segfaults while having negligible impact on performance.  I'm looking into that right now.  It's a tradeoff, and given the requirements of our users, I think anything with a non-negligible performance cost is probably a bad idea.

It's also good to keep a historical perspective in mind.  30 years ago, a bad memory access would often crash your computer.  Protected memory fixed that problem, and instead of letting memory get corrupted the OS just reports a segmentation violation to you and kills the process.  So a segfault is already way better than what used to be standard!  An exception would be even better, and we should aim for that when possible.  But there are much worse things than segfaults.

**Comment by peastman** [comment]
Give #783 a try.

**Comment by swails** [comment]
What happens on the other platforms?  Segfault?  Or catchable exception?

**Comment by peastman** [comment]
Most likely, neither.  The simulation will just keep running with all your coordinates sitting at nan.  The reporters check for that, though, so you'll get an exception the next time one of those runs.

**Comment by jchodera** [comment]
It would be convenient if the other platforms did some sort of `NaN` detection in the integration loop, but given that we can check for `NaN`s after `step()` returns (since there is no segfault), this would only be for convenience.  Maybe something to keep in mind for the future?

**Comment by peastman** [comment]
The problem is that it would require GPU to CPU communication.  We don't want to do that except when absolutely necessary.

**Comment by swails** [comment]
> The problem is that it would require GPU to CPU communication. We don't want to do that except when absolutely necessary.

Why's that?  Why not check for `NaN` directly on the GPU?

**Comment by peastman** [comment]
And then what?  We can't throw an exception from the GPU.

**Comment by jlmaccal** [comment]
I think you can _assert, which returns with an error code, which then throws an OpenMMException. There is also _isfinite to check for NaN.

**Comment by jchodera** [comment]
Did we ever find a good resolution to the `segfault` issue?

**Comment by jchodera** [comment]
Ah, this was indeed addressed by #783!
